src/promisify.js

```diff
@@ -0,0 +1,142 @@
+ *   more than one value - by default, promisify will only resolve the promise with the first value.
+ *     false    - only resolve the promise with the first value, default behavior
+ *     true     - resolve the promise with an array containing the varadic arguments
+ *     String[] - the names of the arguments, which will be used to create an object of values.
+ * @return {Function: Promise} The promisified function.
+ */
+function promisify(orig, {varadic} = {}) {
+  if (typeof orig !== 'function') {
+    throw new TypeError('promisify requires a function');
+  }
+
+  if (orig[kCustomPromisifiedSymbol]) {
+    const fn = orig[kCustomPromisifiedSymbol];
+    if (typeof fn !== 'function') {
+      throw new TypeError('The [util.promisify.custom] property must be a function');
+    }
+    Object.defineProperty(fn, kCustomPromisifiedSymbol, {
+      value: fn, enumerable: false, writable: false, configurable: true
+    });
+    return fn;
+  }
+
+  function fn(...args) {
+    return new Promise((resolve, reject) => {
+      try {
+        orig.call(this, ...args, (err, ...values) => {
+          if (err) {
+            reject(err);
+          } else if (Array.isArray(varadic)) {
+            const obj = {};
+            for (let i = 0; i < varadic.length; i++) {
+              obj[varadic[i]] = values[i];
+            }
+            resolve(obj);
+          } else {
+            resolve(varadic ? values : values[0]);
+          }
+        });
+      } catch (err) {
+        reject(err);
+      }
+    });
+  }
+
+  Object.setPrototypeOf(fn, Object.getPrototypeOf(orig));
+
+  Object.defineProperty(fn, kCustomPromisifiedSymbol, {
+    value: fn, enumerable: false, writable: false, configurable: true
+  });
+  return Object.defineProperties(fn, Object.getOwnPropertyDescriptors(orig));
+}
+
+/**
+ * Promisify the given names on the given object, and create a copy of the object.
+ *
+ * @param {*} obj A value that can have properties.
+ * @param {String[]} methodNames The methods to promisify.
+ * @param {Boolean|String[]} options.varadic See the documentation for promisify.
+ * @return {Object} The promisified object.
+ */
+function promisifyMethods(obj, methodNames, options = {}) {
+  if (!obj) {
+    // This object could be anything, including a function, a real object, or an array.
+    throw new TypeError('promisify.methods requires a truthy value');
+  }
+
+  const out = {};
+
+  for (let methodName of methodNames) {
+    out[methodName] = promisify(obj[methodName].bind(obj), options);
+  }
+
+  return out;
+}
+
+/**
+ * Promisify all functions on the given object.
+ *
+ * @param {*} obj A value that can have properties.
+ * @param {Boolean} options.copyAll Whether to copy the iterable, non-method properties to the new
+ *   object.
+ * @param {Boolean|String[]} options.varadic See the documentation for promisify.
+ * @return {Object} The promisified object.
+ */
+function promisifyAll(obj, options = {}) {
```

Comment:
Well, maybe. The hope is that the API is clear enough (maybe through docs) that it's understood that the returned object isn't intended to be a "replica" of the original.